Improvements to the TTTrack_TrackWord

aperloff opened the pull request:
#### PR description:

This PR fixes an issue with the track parameter digitization. It also converts the incoming floating point phi value from global to sector local coordinates during construction and the setting of the track word parameters. These are the changes which were discussed in https://github.com/cms-l1t-offline/cmssw/pull/896#discussion_r636079432 and https://github.com/cms-l1t-offline/cmssw/pull/938.

#### PR validation:

It was checked that these changes will compile and that they pass the clang-format and clang-tidy checks. So far, no other checks have been performed.

@tomalin @gpetruc @skinnari @emacdonald16

aperloff commented:
@tomalin @gpetruc Can you please comment on these proposed changes?

tomalin commented:
TTTrack_TrackWord::localPhi() looks unnecessarily complicated. Can't you replace it with something like:

#include "DataFormats/Math/interface/deltaPhi.h"
localPhi = reco::deltaPhi(globalPhi - (sector * sectorWidth);

Though please check this gives correct answer.

tomalin commented:
The implementation of TTTrackWord_TrackWord::digitizeSignedValue() looks OK to me. Though to make more obvious what is going on, I'd suggest starting with:

// Digitize
int digi_value = std::floor(value / lsb)
// Do two's compliment encoding (and recover values slightly outside digi range).
(perhaps use twos_value in this function to refer to two's compliment value?)

tomalin commented:
TTTrackWord_TrackWord::unpackSignedValue(), to match naming convention of previous function should be undigitizeSignedValue(). 

Please make clear in this function, via comments & variable names, which digitized values are two's compliment (e.g. the input "bits" argument") and which are not. e.g. Rename "bits" to "twos_value".

Suggest starting function with:

// Select only bits in specified range. (N.B. Is this really necessary? Are there ever bits outside the range?)
unsigned int digitized_maximum = (1 << nBits) - 1;
bits = bits & digitized_maximum;

The maths for -ve numbers looks complicated (and may be wrong). I'd have expected something simpler:
// Convert from twos compliment to standard digi.
int digi_value = int(bits);
if (bits & (1 << (nBits - 1))) {  // check if -ve number
  int digi_value -= (1 << nBits);
}
return (digi_value + 0.5)*lsb;

If I set nBits=3, lsb=1, bits=4 (a twos compliment number), then looking at https://en.wikipedia.org/wiki/Two%27s_complement , this should be equivalent to a normal digitized value of -4, and hence to a floating point value of -3.5. But I think the code in unpackSignedValue() would yield -4.5.

aperloff commented:
> (N.B. Is this really necessary? Are there ever bits outside the range?)
<details>
<summary>Expand full quote...</summary>
<p>

```c++
// Select only bits in specified range. (N.B. Is this really necessary? Are there ever bits outside the range?)
unsigned int digitized_maximum = (1 << nBits) - 1;
bits = bits & digitized_maximum;
```

</p>
</details>

I do think the range selection is a good idea. If we don't make sure that the msb is at `1 << (nBits -1)` then the check for the sign won't be using the correct bit. We could change this to throw an error if the `bits` unsigned int argument contains a value with more bits than nBits. Would you prefer that?

aperloff commented:
Hi @tomalin, I believe I have addressed all of your comments with these latest commits. I have confirmed that the `digitizeSignedValue` and `undigitizeSignedValue` functions work as one would expect (see https://github.com/cms-l1t-offline/cmssw/pull/940/files#diff-6e795a8c5ee6eae0757f2898b146a5f1614067afedca66b20785518e915e4891R193-R205) and I have also simplified the `localPhi` function. Let me know if you have any additional comments.

tomalin commented on DataFormats/L1TrackTrigger/interface/TTTrack_TrackWord.h:
```diff
@@ -257,16 +263,50 @@ class TTTrack_TrackWord {
+  unsigned int countSetBitsInGivenRange(unsigned int value, unsigned int lsb, unsigned int msb) const {
+    // Calculate a number 'num' having 'msb' number of bits and bits in the range 'lsb' to 'msb' are the only set bits
+    // Note: 'lsb' and 'msb' are ordered as follows 1 <= lsb <= msb <= <number of bits in value>
+    int num = ((1 << msb) - 1) ^ ((1 << (lsb - 1)) - 1);
```
Perhaps clearer if replace "int num" by "unsigned int mask"

aperloff replied:
Changed as recommended.

tomalin commented on DataFormats/L1TrackTrigger/interface/TTTrack_TrackWord.h:
```diff
@@ -257,16 +263,50 @@ class TTTrack_TrackWord {
+
+    // Calculate the maxmum possible positive value given an output of nBits in size
+    unsigned int digitizedMaximum = (1 << (nBits - 1)) - 1;  // The remove 1 bit from nBits to account for the sign
+    unsigned int digitizedMinimum = (1 << (nBits - 1));
```
or simplier: digitizedMinimum = digitizedMaximum + 1;

aperloff replied:
Changed as recommended.

tomalin commented on DataFormats/L1TrackTrigger/interface/TTTrack_TrackWord.h:
```diff
@@ -257,16 +263,50 @@ class TTTrack_TrackWord {
+    unsigned int digitizedMaximum = (1 << (nBits - 1)) - 1;  // The remove 1 bit from nBits to account for the sign
+    unsigned int digitizedMinimum = (1 << (nBits - 1));
+
+    // Saturate the digitized value
```
If the goal of this is to rescue numbers that are slightly outside the digi range, I don't think it works. If digi range is from -4 to +3 (lsb=1,nbits=3), then it digitizedMinimum=4, so this code will only change the value of twosValue of digitizedValue (which is abs of digi value) is 5 or more. So if numerical errors had caused the input digi to be +4, it would still lie outside the allowed range after this code.

aperloff replied:
Hi @tomalin, I see what you mean. What do you think about changing the conditional to:
```
    if ( (value >= 0) && (digitizedValue > digitizedMaximum) ) {
      twosValue = digitizedMaximum;
    }
    else if( (value < 0) && (digitizedValue > digitizedMinimum) ) {
      twosValue = digitizedMinimum;
    }
```
Using the input values of `[3,4,5,-4,-5]` with `lsb=1` and `nBits=3` I find the correct behavior:
```
value ==> twosValue
3 ==> 3 // neither conditional is true
4 ==> 3 // 1st conditional evaluates to true
5 ==> 3 // 1st conditional evaluates to true
-4 ==> 4 // neither conditional is true
-5 ==> 4 // 2nd conditional evaluates to true
```
Do you agree that this would satisfy the issue?

tomalin replied:
I agree that would be correct. Though if the range checking is now asymmetric, I'd suggest the code would be clearer by removing the abs() from https://github.com/cms-l1t-offline/cmssw/blob/e4bbee047752a8b5f9ad1bdbc6d7c78950b6cb68/DataFormats/L1TrackTrigger/interface/TTTrack_TrackWord.h#L300 , so you work with conventional digi values . i.e.

int digitizedValue = std::floor(value / lsb);
int digitizedMaximum = (1 << (nBits - 1)) - 1;  
int digitizedMinimum = -(digitizedMaximum + 1);
if (digitizedValue > digitizedMaximum) digitizedValue = digitizedMaximum;
if (digitizedValue < digitizedMinimum) digitizedValue = digitizedMinimum;
int twosValue;
if (digitizedValue >= 0) {
   twosValue = digitizedValue;
} else {
  twosValue = digitizedValue + (1 << nBits);
}

gpetruc replied:
I second @tomalin's suggestion.

Depending on taste, you could replace the two if's with std::min and std::max, or better if c++17 is allowed in CMSSW  then [std::clamp](https://en.cppreference.com/w/cpp/algorithm/clamp), i.e. `digitizedValue = std::clamp(digitizedValue, digitizedMinimum, digitizedMaximum)`

aperloff replied:
I have integrated both of your suggestions.

tomalin commented on DataFormats/L1TrackTrigger/interface/TTTrack_TrackWord.h:
```diff
@@ -268,14 +323,23 @@ class TTTrack_TrackWord {
-      isign = -1;
-      bits = (1 << (nBits + 1)) - bits;  // if negative, flip everything for two's complement encoding
+  double undigitizeSignedValue(unsigned int twosValue, unsigned int nBits, double lsb) const {
+    // Check that none of the bits above the nBits-1 bit, in a range of [0, nBits-1], are set.
```
I'm not sure why this check is needed. But can't it be done in a single line of code, eliminating the need for the countSetBits* functions? e.g.:
assert (twosValue >> (lsb + nbits) == 0);

gpetruc replied:
I think it should be `assert((twosValue >> nbits) == 0)`

aperloff replied:
The check is now done with the line suggested by @gpetruc.

rekovic commented:
@aperloff 
It looks like the discussion here is still ongoing and needs to converge before we merge this.
Please also fix the conflicts now after #938 has been merged.

aperloff commented:
@tomalin @gpetruc All of the review comments have been addressed. I've also rebased in light of the recent merges to the `phase2-l1t-integration-CMSSW_11_1_7` branch and re-run the code checks and code format checks. Can you please take another look?

tomalin commented on DataFormats/L1TrackTrigger/interface/TTTrack_TrackWord.h:
```diff
@@ -269,14 +302,19 @@ class TTTrack_TrackWord {
+    //  any bigger than ((1 << nBits) - 1).
+    assert((twosValue >> nBits) == 0);
+
+    // Convert from twos compliment to floating point values
```
I'm not sure what a non-twos-complement digitized number is called. But I'd suggest comment at L311 should be "Convert from twos compliment to normal digitized value", and then before L317, you need comment "Convert to floating point value".

aperloff replied:
Fixed

tomalin reviewed (commented):
The code looks good to me now. Have you validated the digitization? e.g. By starting with the digi values, then undigitizing & redigitizing them, and checking you get back the original numbers?

aperloff commented:
@tomalin Yes, I have validated the digitization. I implemented a "unit test" for the `TTTrack_TrackWord`. You can run it using:
```bash
cd DataFormats/L1TrackTrigger/test/
scram b unittests
```
This returns the following output:
```
%MSG-i TTTrack_TrackWord:  TTTrackTrackWordDummyOneAnalyzer:dummyAna  05-Aug-2021 08:48:11 CDT Run: 1 Event: 1
testDigitizationScheme: (nBits, lsb) = (3, 1)
Floating point value = -4	Digitized value = 4	Un-digitized value = -3.5	Re-digitized value = 4
%MSG
%MSG-i TTTrack_TrackWord:  TTTrackTrackWordDummyOneAnalyzer:dummyAna  05-Aug-2021 08:48:11 CDT Run: 1 Event: 1
testDigitizationScheme: (nBits, lsb) = (3, 1)
Floating point value = 3	Digitized value = 3	Un-digitized value = 3.5	Re-digitized value = 3
%MSG
%MSG-i TTTrack_TrackWord:  TTTrackTrackWordDummyOneAnalyzer:dummyAna  05-Aug-2021 08:48:11 CDT Run: 1 Event: 1
testDigitizationScheme: (nBits, lsb) = (4, 0.5)
Floating point value = -3.5	Digitized value = 9	Un-digitized value = -3.25	Re-digitized value = 9
%MSG
%MSG-i TTTrack_TrackWord:  TTTrackTrackWordDummyOneAnalyzer:dummyAna  05-Aug-2021 08:48:11 CDT Run: 1 Event: 1
testDigitizationScheme: (nBits, lsb) = (4, 0.5)
Floating point value = 3.5	Digitized value = 7	Un-digitized value = 3.75	Re-digitized value = 7
%MSG
```
This indicates to me that the digitization and un-digitization are working.
